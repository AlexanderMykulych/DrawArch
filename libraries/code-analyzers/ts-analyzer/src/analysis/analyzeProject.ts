import { ProgramContainer } from "./types";
import { analyzeSourceFile } from "./analyzeSourceFile";
import { getProgramAndChecker } from "./getProgramAndChecker";


export const analyzeProject = (projPath: string) => {
  const container = getProgramAndChecker(projPath) as ProgramContainer;
  if (container.program) {
    container
      .program
      .getSourceFiles()
      .forEach(sourceFile => analyzeSourceFile({ sourceFile, checker: container.checker }));
  }
  return null;
};
