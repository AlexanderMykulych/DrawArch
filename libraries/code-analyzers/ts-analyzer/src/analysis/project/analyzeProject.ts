import { Project } from 'ts-morph'
import { processors } from '../processors'
import { analyzeSourceFile } from '../source-file/analyzeSourceFile'
import type { CodeInfo, FileProcessor, FileText } from '../types'
import { deduplicate } from './deduplicate'
import { initializeProject } from './initializeProject'

export async function analyzeProject(projPath: string): Promise<CodeInfo[]> {

  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      allowJs: true,
      rootDir: projPath,
    },
    skipLoadingLibFiles: false,
  })

  if (await initializeProject(projPath, project, 'custom')) {
    const codeInfos = project
      .getSourceFiles()
      .flatMap(x => analyzeSourceFile(x))

    return deduplicate(codeInfos)
  }
  return []
}

export async function processFile(file: FileText) {
  let resultFile: FileText = file
  let notUsedProcessors = [
    ...processors,
  ]
  let processor: FileProcessor | undefined

  do {
    processor = notUsedProcessors.find(x => x.isMatch(file))

    if (processor) {
      notUsedProcessors = notUsedProcessors.filter(x => x !== processor)
      resultFile = await processor.process(file)
    }
  } while (processor)

  return resultFile
}
