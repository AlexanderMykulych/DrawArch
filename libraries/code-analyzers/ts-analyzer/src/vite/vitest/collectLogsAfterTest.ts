import { ActionEntity, ActionEntityType, CodeInfo, CodeInfoType, Relationship, RelationType } from "../../analysis/types"
import { useNeo4j } from "../../neo4j/useNeo4j"
import type { LogItem, StartLogItem } from "../shim"


export async function collectLogsAfterTest(insertToDb: true): Promise<void>
export async function collectLogsAfterTest(): Promise<CodeInfo[]>
export async function collectLogsAfterTest(insertToDb?: true): Promise<CodeInfo[] | void> {
  if (!globalThis.__getLogs) {
    return
  }

  const logs = globalThis.__getLogs()

  const codeInfos = logsToCodeInfo(logs)

  if (insertToDb) {
    const db = useNeo4j()
    await db.insertCodeInfos(codeInfos);
  } else {
    return codeInfos
  }
}


export function logsToCodeInfo(logs: LogItem[]): CodeInfo[] {
  return logs
    .flatMap<CodeInfo>((item, index, items) => {
      const result = []
      if (item.type === 'end') {
        const callEntity: ActionEntity = {
          type: CodeInfoType.Entity,
          entityType: ActionEntityType.Call,
          id: getCallActionId(item, index),
          executionTime: item.execTime,
        }
        const callRelation: Relationship = {
          type: CodeInfoType.Relationship,
          relationType: RelationType.CallEntity,
          from: callEntity.id,
          to: item.id,
        }

        result.push(callEntity, callRelation)

        const [parentStartItem, parentIndex] = findParentStartLogFromIndex(items, index, item.id)

        if (parentStartItem) {
          const parentRelation: Relationship = {
            type: CodeInfoType.Relationship,
            relationType: RelationType.Parent,
            from: getCallActionId(parentStartItem, parentIndex),
            to: callEntity.id,
          }
          result.push(parentRelation)
        }
      }
      return result
    })
}

export function findStartLogFromIndex(items: LogItem[], index: number, id: string): StartLogItem {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id === id) {
      return startItem
    }
  }
  throw 'Can not find start element for: ' + id
}

export function findParentStartLogFromIndex(items: LogItem[], index: number, id: string): [StartLogItem, number] | [null, null] {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id !== id) {
      return [startItem, i]
    }
  }

  return [null, null]
}

export function getCallActionId(item: LogItem, index: number): string {
  // return `call:${item.id}`
  return `call[${index}]:${item.id}`
}