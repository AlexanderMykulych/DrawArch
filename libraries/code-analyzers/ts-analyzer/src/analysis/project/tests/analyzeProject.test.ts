import path from 'path'
import { expect, test } from 'vitest'
import { analyzeProject } from '../analyzeProject'

test('should analyze simple ts project', async () => {
  const result = await analyzeProject(path.resolve(__dirname, './cases/project4'))

  expect(result).toMatchSnapshot()
})

test('should analyze vue-ts project', async () => {
  const result = await analyzeProject(path.resolve(__dirname, './cases/project5'))

  expect(result).toMatchSnapshot()

  const expObj = result
    .flatMap(x => Object.entries(x) as [string, string][])
    .filter(([_, val]) => val.includes ? val.includes('[object Object]') : false)
  expect(expObj,
    `shouldn't contain [object Object]. ${JSON.stringify(expObj)}`,
  ).empty

  expect(result).toEqual(expect.arrayContaining([
    expect.objectContaining({
      id: '/cmp.vue.ts',
      type: 'entity',
    }),
    expect.objectContaining({
      id: '/util.ts/fn1',
      type: 'entity',
    }),
    expect.objectContaining({
      to: '/util.ts/fn1',
    }),
  ]))
})

test('should analyze complex vue-ts project', async () => {
  const result = await analyzeProject(path.resolve(__dirname, './cases/project6'))

  expect(result).toMatchSnapshot()

  const expObj = result
    .flatMap(x => Object.entries(x) as [string, string][])
    .filter(([_, val]) => val.includes ? val.includes('[object Object]') : false)
  expect(expObj,
    `shouldn't contain [object Object]. ${JSON.stringify(expObj)}`,
  ).empty

  expect(result).toEqual(expect.arrayContaining([
    expect.objectContaining({
      id: '/service.ts/getService/method1',
      type: 'entity',
    }),
    expect.objectContaining({
      id: '/util.ts/fn1',
      type: 'entity',
    }),
    expect.objectContaining({
      from: '/util.ts/fn1',
      to: '/service.ts/getService/method1',
    }),
  ]))
})
