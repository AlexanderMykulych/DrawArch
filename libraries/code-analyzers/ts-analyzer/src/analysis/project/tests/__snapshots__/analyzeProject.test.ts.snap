// Vitest Snapshot v1

exports[`should analyze complex vue-ts project 1`] = `
[
  {
    "from": "/cmp.vue.ts",
    "relationType": "import",
    "to": "/utils/util.ts",
    "type": "relationship",
  },
  {
    "entityType": "file",
    "filePath": "/cmp.vue.ts",
    "id": "/cmp.vue.ts",
    "type": "entity",
  },
  {
    "entityType": "file",
    "filePath": "/utils/util.ts",
    "id": "/utils/util.ts",
    "type": "entity",
  },
  {
    "from": "/cmp.vue.ts",
    "relationType": "import",
    "to": "vue",
    "type": "relationship",
  },
  {
    "entityType": "import_declaration",
    "filePath": "<unknow>",
    "id": "vue",
    "type": "entity",
  },
  {
    "entityType": "function",
    "filePath": "/cmp.vue.ts",
    "id": "/cmp.vue.ts/render",
    "type": "entity",
    "typeString": "typeof import(\\"/cmp.vue\\").render",
  },
  {
    "entityType": "variable",
    "filePath": "<unknown>",
    "id": "/cmp.vue.ts/openBlock",
    "type": "entity",
  },
  {
    "entityType": "variable",
    "filePath": "<unknown>",
    "id": "/cmp.vue.ts/createElementBlock",
    "type": "entity",
  },
  {
    "entityType": "variable",
    "filePath": "<unknown>",
    "id": "/cmp.vue.ts/toDisplayString",
    "type": "entity",
  },
  {
    "from": "/cmp.vue.ts/render",
    "relationType": "use",
    "to": "/cmp.vue.ts/openBlock",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts/render",
    "relationType": "use",
    "to": "/cmp.vue.ts/createElementBlock",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts/render",
    "relationType": "use",
    "to": "/cmp.vue.ts/toDisplayString",
    "type": "relationship",
  },
  {
    "entityType": "function",
    "filePath": "/cmp.vue.ts",
    "id": "/cmp.vue.ts/setup",
    "type": "entity",
    "typeString": "(__props: any, { expose }: { expose: any; }) => { val: any; fn1: typeof import(\\"/utils/util\\").fn1; }",
  },
  {
    "entityType": "function",
    "filePath": "/utils/util.ts",
    "id": "/utils/util.ts/fn1",
    "type": "entity",
    "typeString": "typeof import(\\"/utils/util\\").fn1",
  },
  {
    "entityType": "variable",
    "filePath": "/node_modules/typescript/lib/lib.es5.d.ts",
    "id": "/node_modules/typescript/lib/lib.es5.d.ts/Object",
    "type": "entity",
  },
  {
    "entityType": "function",
    "filePath": "/node_modules/typescript/lib/lib.es5.d.ts",
    "id": "/node_modules/typescript/lib/lib.es5.d.ts/defineProperty",
    "type": "entity",
    "typeString": "<T>(o: T, p: PropertyKey, attributes: PropertyDescriptor & ThisType<any>) => T",
  },
  {
    "from": "/cmp.vue.ts/setup",
    "relationType": "use",
    "to": "/utils/util.ts/fn1",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts/setup",
    "relationType": "use",
    "to": "/node_modules/typescript/lib/lib.es5.d.ts/Object",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts/setup",
    "relationType": "use",
    "to": "/node_modules/typescript/lib/lib.es5.d.ts/defineProperty",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts",
    "relationType": "parent",
    "to": "/cmp.vue.ts/render",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts",
    "relationType": "parent",
    "to": "/cmp.vue.ts/setup",
    "type": "relationship",
  },
  {
    "from": "/utils/util.ts",
    "relationType": "parent",
    "to": "/utils/util.ts/fn1",
    "type": "relationship",
  },
  {
    "entityType": "file",
    "filePath": "/node_modules/typescript/lib/lib.es5.d.ts",
    "id": "/node_modules/typescript/lib/lib.es5.d.ts",
    "type": "entity",
  },
  {
    "from": "/node_modules/typescript/lib/lib.es5.d.ts",
    "relationType": "parent",
    "to": "/node_modules/typescript/lib/lib.es5.d.ts/Object",
    "type": "relationship",
  },
  {
    "from": "/node_modules/typescript/lib/lib.es5.d.ts",
    "relationType": "parent",
    "to": "/node_modules/typescript/lib/lib.es5.d.ts/defineProperty",
    "type": "relationship",
  },
  {
    "from": "/index.ts",
    "relationType": "import",
    "to": "/cmp.vue.ts",
    "type": "relationship",
  },
  {
    "entityType": "file",
    "filePath": "/index.ts",
    "id": "/index.ts",
    "type": "entity",
  },
  {
    "entityType": "file",
    "filePath": "/package.json",
    "id": "/package.json",
    "type": "entity",
  },
  {
    "entityType": "function",
    "filePath": "/services/service.ts",
    "id": "/services/service.ts/getService",
    "type": "entity",
    "typeString": "typeof import(\\"/services/service\\").getService",
  },
  {
    "entityType": "function",
    "filePath": "/services/service.ts",
    "id": "/services/service.ts/getService/method1",
    "type": "entity",
    "typeString": "(a: number, b: number) => number",
  },
  {
    "entityType": "function",
    "filePath": "/services/service.ts",
    "id": "/services/service.ts/getService/prop1/method2",
    "type": "entity",
    "typeString": "(a: number) => number",
  },
  {
    "entityType": "file",
    "filePath": "/services/service.ts",
    "id": "/services/service.ts",
    "type": "entity",
  },
  {
    "from": "/services/service.ts",
    "relationType": "parent",
    "to": "/services/service.ts/getService",
    "type": "relationship",
  },
  {
    "from": "/services/service.ts",
    "relationType": "parent",
    "to": "/services/service.ts/getService/method1",
    "type": "relationship",
  },
  {
    "from": "/services/service.ts/getService",
    "relationType": "parent",
    "to": "/services/service.ts/getService/method1",
    "type": "relationship",
  },
  {
    "entityType": "property",
    "filePath": "/services/service.ts",
    "id": "/services/service.ts/getService/prop1",
    "type": "entity",
  },
  {
    "from": "/services/service.ts",
    "relationType": "parent",
    "to": "/services/service.ts/getService/prop1/method2",
    "type": "relationship",
  },
  {
    "from": "/services/service.ts/getService",
    "relationType": "parent",
    "to": "/services/service.ts/getService/prop1/method2",
    "type": "relationship",
  },
  {
    "from": "/services/service.ts/getService/prop1",
    "relationType": "parent",
    "to": "/services/service.ts/getService/prop1/method2",
    "type": "relationship",
  },
  {
    "from": "/utils/util.ts",
    "relationType": "import",
    "to": "/services/service.ts",
    "type": "relationship",
  },
  {
    "from": "/utils/util.ts/fn1",
    "relationType": "use",
    "to": "/services/service.ts/getService",
    "type": "relationship",
  },
  {
    "from": "/utils/util.ts/fn1",
    "relationType": "use",
    "to": "/services/service.ts/getService/method1",
    "type": "relationship",
  },
  {
    "from": "/utils/util.ts/fn1",
    "relationType": "use",
    "to": "/services/service.ts/getService/prop1",
    "type": "relationship",
  },
  {
    "from": "/utils/util.ts/fn1",
    "relationType": "use",
    "to": "/services/service.ts/getService/prop1/method2",
    "type": "relationship",
  },
  {
    "from": "/services/service.ts",
    "relationType": "parent",
    "to": "/services/service.ts/getService/prop1",
    "type": "relationship",
  },
  {
    "from": "/services/service.ts/getService",
    "relationType": "parent",
    "to": "/services/service.ts/getService/prop1",
    "type": "relationship",
  },
]
`;

exports[`should analyze simple ts project 1`] = `
[
  {
    "from": "/index.ts",
    "relationType": "import",
    "to": "/util.ts",
    "type": "relationship",
  },
  {
    "entityType": "file",
    "filePath": "/index.ts",
    "id": "/index.ts",
    "type": "entity",
  },
  {
    "entityType": "file",
    "filePath": "/util.ts",
    "id": "/util.ts",
    "type": "entity",
  },
  {
    "entityType": "function",
    "filePath": "/util.ts",
    "id": "/util.ts/fn1",
    "type": "entity",
    "typeString": "(a: number) => number",
  },
  {
    "from": "/util.ts",
    "relationType": "parent",
    "to": "/util.ts/fn1",
    "type": "relationship",
  },
]
`;

exports[`should analyze vue-ts project 1`] = `
[
  {
    "from": "/cmp.vue.ts",
    "relationType": "import",
    "to": "/util.ts",
    "type": "relationship",
  },
  {
    "entityType": "file",
    "filePath": "/cmp.vue.ts",
    "id": "/cmp.vue.ts",
    "type": "entity",
  },
  {
    "entityType": "file",
    "filePath": "/util.ts",
    "id": "/util.ts",
    "type": "entity",
  },
  {
    "from": "/cmp.vue.ts",
    "relationType": "import",
    "to": "vue",
    "type": "relationship",
  },
  {
    "entityType": "import_declaration",
    "filePath": "<unknow>",
    "id": "vue",
    "type": "entity",
  },
  {
    "entityType": "function",
    "filePath": "/cmp.vue.ts",
    "id": "/cmp.vue.ts/render",
    "type": "entity",
    "typeString": "typeof import(\\"/cmp.vue\\").render",
  },
  {
    "entityType": "variable",
    "filePath": "<unknown>",
    "id": "/cmp.vue.ts/openBlock",
    "type": "entity",
  },
  {
    "entityType": "variable",
    "filePath": "<unknown>",
    "id": "/cmp.vue.ts/createElementBlock",
    "type": "entity",
  },
  {
    "entityType": "variable",
    "filePath": "<unknown>",
    "id": "/cmp.vue.ts/toDisplayString",
    "type": "entity",
  },
  {
    "from": "/cmp.vue.ts/render",
    "relationType": "use",
    "to": "/cmp.vue.ts/openBlock",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts/render",
    "relationType": "use",
    "to": "/cmp.vue.ts/createElementBlock",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts/render",
    "relationType": "use",
    "to": "/cmp.vue.ts/toDisplayString",
    "type": "relationship",
  },
  {
    "entityType": "function",
    "filePath": "/cmp.vue.ts",
    "id": "/cmp.vue.ts/setup",
    "type": "entity",
    "typeString": "(__props: any, { expose }: { expose: any; }) => { val: any; fn1: typeof import(\\"/util\\").fn1; }",
  },
  {
    "entityType": "function",
    "filePath": "/util.ts",
    "id": "/util.ts/fn1",
    "type": "entity",
    "typeString": "typeof import(\\"/util\\").fn1",
  },
  {
    "entityType": "variable",
    "filePath": "/node_modules/typescript/lib/lib.es5.d.ts",
    "id": "/node_modules/typescript/lib/lib.es5.d.ts/Object",
    "type": "entity",
  },
  {
    "entityType": "function",
    "filePath": "/node_modules/typescript/lib/lib.es5.d.ts",
    "id": "/node_modules/typescript/lib/lib.es5.d.ts/defineProperty",
    "type": "entity",
    "typeString": "<T>(o: T, p: PropertyKey, attributes: PropertyDescriptor & ThisType<any>) => T",
  },
  {
    "from": "/cmp.vue.ts/setup",
    "relationType": "use",
    "to": "/util.ts/fn1",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts/setup",
    "relationType": "use",
    "to": "/node_modules/typescript/lib/lib.es5.d.ts/Object",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts/setup",
    "relationType": "use",
    "to": "/node_modules/typescript/lib/lib.es5.d.ts/defineProperty",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts",
    "relationType": "parent",
    "to": "/cmp.vue.ts/render",
    "type": "relationship",
  },
  {
    "from": "/cmp.vue.ts",
    "relationType": "parent",
    "to": "/cmp.vue.ts/setup",
    "type": "relationship",
  },
  {
    "from": "/util.ts",
    "relationType": "parent",
    "to": "/util.ts/fn1",
    "type": "relationship",
  },
  {
    "entityType": "file",
    "filePath": "/node_modules/typescript/lib/lib.es5.d.ts",
    "id": "/node_modules/typescript/lib/lib.es5.d.ts",
    "type": "entity",
  },
  {
    "from": "/node_modules/typescript/lib/lib.es5.d.ts",
    "relationType": "parent",
    "to": "/node_modules/typescript/lib/lib.es5.d.ts/Object",
    "type": "relationship",
  },
  {
    "from": "/node_modules/typescript/lib/lib.es5.d.ts",
    "relationType": "parent",
    "to": "/node_modules/typescript/lib/lib.es5.d.ts/defineProperty",
    "type": "relationship",
  },
  {
    "from": "/index.ts",
    "relationType": "import",
    "to": "/cmp.vue.ts",
    "type": "relationship",
  },
  {
    "entityType": "file",
    "filePath": "/index.ts",
    "id": "/index.ts",
    "type": "entity",
  },
]
`;
